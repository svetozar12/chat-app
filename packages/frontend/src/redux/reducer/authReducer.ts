import { Action } from "../state";
import { ActionType } from "../state";
import { InitialState } from "../state";
import { HYDRATE } from "next-redux-wrapper";

const initialState = {
  remember_me: false,
  input: "",
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: "",
};
const reducer = (state: InitialState = initialState, action: Action) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.SAVE_INPUT:
      return { ...state, input: action.payload };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    case ActionType.SIGN_OUT:
      return { state: undefined };
    default:
      return state;
  }
};

export default reducer;
