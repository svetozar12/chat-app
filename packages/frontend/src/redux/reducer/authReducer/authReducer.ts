import { HYDRATE } from "next-redux-wrapper";
import { IAuthState } from "./state";
import { ActionType } from "../../types";
export const initialState: IAuthState = {
  remember_me: false,
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: {
    username: "",
    id: "",
    token: "",
    refresh_token: "",
  },
};
const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default reducer;
export { initialState as AuthState };
