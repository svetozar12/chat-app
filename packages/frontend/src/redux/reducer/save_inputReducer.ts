import { Action, ActionType, InitialState3 } from "../state";
import { HYDRATE } from "next-redux-wrapper";
export const initialState = {
  input_username: "",
  input_password: "",
  input_email: "",
  input_gender: "",
  notification_number: 0,
};
const saveInputReducer = (state: InitialState3 = initialState, action: Action) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.SAVE_INPUT_USERNAME:
      return { ...state, input_username: action.payload };
    case ActionType.SAVE_INPUT_PASSWORD:
      return { ...state, input_password: action.payload };
    case ActionType.SAVE_INPUT_EMAIL:
      return { ...state, input_email: action.payload };
    case ActionType.SAVE_INPUT_GENDER:
      return { ...state, input_gender: action.payload };
    case ActionType.NOTIFICATION_NUMBER:
      return { ...state, notification_number: action.payload };
    default:
      return state;
  }
};

export default saveInputReducer;
