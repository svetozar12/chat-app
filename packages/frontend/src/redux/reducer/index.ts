import { combineReducers } from "redux";
import reducer from "./authReducer";
import setReducer from "./setReducer";
import saveInputReducer from "./save_inputReducer";
import messageReducer from "./messageReducer";

const reducers = combineReducers({
  authReducer: reducer,
  setReducer,
  messageReducer,
  saveInputReducer,
});

export default reducers;
