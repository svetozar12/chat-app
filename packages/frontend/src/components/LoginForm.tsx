import { useSelector, useDispatch } from "react-redux";
import Link from "next/dist/client/link";
import { InitialState } from "../redux/state";

function LoginForm({
  handleSubmit,
}: {
  handleSubmit: (e: React.MouseEvent<HTMLButtonElement>) => Promise<void>;
}) {
  const state = useSelector(
    (state: { authReducer: InitialState }) => state.authReducer,
  );
  const dispatch = useDispatch();
  return (
    <div className="login-box">
      <form style={{ height: "100vh" }} className="container">
        <h1 className="alert" style={{ color: "red" }}>
          {state.bad}
        </h1>
        <h1>Login</h1>
        <input
          value={state.input_username}
          onChange={(e) =>
            dispatch({ type: "SAVE_INPUT_USERNAME", payload: e.target.value })
          }
          type="text"
          name="username"
          placeholder="username ..."
        />
        <input
          value={state.input_password}
          onChange={(e) =>
            dispatch({ type: "SAVE_INPUT_PASSWORD", payload: e.target.value })
          }
          type="password"
          name="password"
          placeholder="password ..."
        />
        <div className="clickable">
          <Link href="/register">
            <a className="link" style={{ color: "var(--main-blue)" }}>
              Create an account.
            </a>
          </Link>
          <label htmlFor="checkbox">
            {" "}
            <input
              type="checkbox"
              id="checkbox"
              onChange={(e) =>
                dispatch({
                  type: "REMEMBER_ME_CHECK",
                  payload: e.target.checked,
                })
              }
              style={{ width: "20px", height: "40px" }}
            />
            Remember me
          </label>
        </div>
        <button onClick={handleSubmit} type="submit">
          login
        </button>
      </form>
    </div>
  );
}

export default LoginForm;
