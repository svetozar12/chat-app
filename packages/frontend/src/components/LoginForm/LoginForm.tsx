import { useSelector, useDispatch } from "react-redux";
import Link from "next/dist/client/link";
import { InitialState, InitialState3 } from "../../redux/state";
import styled from "@emotion/styled";
import { Alerts } from "../Alerts/Alerts";
import { css } from "@emotion/css";
export const Input = styled.input`
  width: 100%;
  height: 2rem;
  margin: 0.5rem 0;
  border: 1px solid var(--input-border-color);
  border-radius: 5px;
  transition: 0.3s;
  padding: 1.3rem 0.9rem;
`;

const CheckBox = styled.input`
  width: 20px;
  height: 40px;
`;

export const Form = styled.form`
  width: 40%;
  height: 50vh;
  padding: 3rem;
  border-radius: 5px;
  background: #fff;
  box-shadow: 0 2px 30px rgba(0, 0, 0, 0.1);
  @media (max-width: 1010px) {
    width: 70%;
  }
`;

export const Button = styled.button`
  margin-top: 1rem;
  border-radius: 5px;
  width: 60%;
  background-color: var(--button-blue);
  color: var(--main-white);
  border: 1px solid var(--input-border-color);
  padding: 1rem;
  cursor: pointer;
  transition: 0.2s;
  font-weight: bold;
  font-size: 1rem;
`;

const Clickable = styled.div`
  display: flex;
  align-items: center;
`;

const Link_anchor = styled.a`
  text-decoration: none;
  cursor: pointer;
  border: 1px solid transparent;
  margin: 1rem;
`;

export const Label_container = styled.div`
  dispay: flex;
  flex-direction: collumn;
  width: 60%;
`;

export const Form_header = styled.h1`
  padding: 1rem;
  height: auto;
  text-align: center;
  margin: 0;
  width: 40%;
  border-top-right-radius: 5px;
  border-top-left-radius: 5px;
  color: var(--main-white);
  background: var(--form-gray);
  @media (max-width: 1010px) {
    width: 70%;
  }
`;

function LoginForm({
  handleSubmit,
}: {
  handleSubmit: (e: React.MouseEvent<HTMLButtonElement>) => Promise<void>;
}) {
  const state = useSelector(
    (state: { authReducer: InitialState }) => state.authReducer,
  );

  const inputState = useSelector(
    (state: { saveInputReducer: InitialState3 }) => state.saveInputReducer,
  );
  const dispatch = useDispatch();

  return (
    <div style={{ height: "100vh", flexDirection: "column" }} className="flex">
      {(state.good || state.bad) && <Alerts />}
      <div style={{ width: "100%", height: "3rem" }}></div>
      <Form_header>Login</Form_header>
      <Form>
        <Label_container>
          <label>Username</label>
          <Input
            value={inputState.input_username}
            onChange={(e) =>
              dispatch({ type: "SAVE_INPUT_USERNAME", payload: e.target.value })
            }
            type="text"
            name="username"
            placeholder="username ..."
          />
        </Label_container>
        <Label_container>
          <label>Password</label>
          <Input
            value={inputState.input_password}
            onChange={(e) =>
              dispatch({ type: "SAVE_INPUT_PASSWORD", payload: e.target.value })
            }
            type="password"
            name="password"
            placeholder="password ..."
          />
        </Label_container>
        <Button onClick={handleSubmit} type="submit">
          Log In
        </Button>
        <Clickable>
          <Link href="/register">
            <Link_anchor
              className="link"
              style={{ color: "var(--button-blue)" }}
            >
              Sign up for chatApp .
            </Link_anchor>
          </Link>
          <CheckBox
            data-testid="checkbox"
            type="checkbox"
            id="checkbox"
            checked={state.remember_me}
            onChange={(e) =>
              dispatch({
                type: "REMEMBER_ME_CHECK",
                payload: e.target.checked,
              })
            }
          />
          <label style={{ marginLeft: ".25rem" }} htmlFor="checkbox">
            {" "}
            Remember me
          </label>
        </Clickable>
      </Form>
    </div>
  );
}

export default LoginForm;
