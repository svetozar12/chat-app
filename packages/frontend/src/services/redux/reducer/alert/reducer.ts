import { AnyAction } from 'redux';
import { IAlert } from 'services/redux/reducer/alert/state';
import { ActionType } from '../../types';

const initialState: IAlert = {
  good: '', // good and bad stand for good alert and bad alert
  bad: '',
};

const alertReducer = (state = initialState, action: AnyAction) => {
  switch (action.type) {
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.payload };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.payload };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.payload };
    default:
      return state;
  }
};

export default alertReducer;
