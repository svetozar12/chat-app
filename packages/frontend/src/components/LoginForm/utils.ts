import { Cookie } from 'next-cookie';
import { NextRouter } from 'next/router';
import { LoginUserMutationVariables, LoginUserMutationFn, LoginUserMutation } from 'services/generated';
import { setAlert } from 'services/redux/reducer/alert/actions';
import { IAuth } from 'services/redux/reducer/auth/state';
import { setInputPassword, setInputUsername } from 'services/redux/reducer/inputs/actions';
import { togglelIsLoading } from 'services/redux/reducer/toggles/actions';

export const handleSubmit = async (
  values: LoginUserMutationVariables,
  auth: IAuth,
  cookie: Cookie,
  router: NextRouter,
  callback: string,
  mutation: { loginUserMutation: LoginUserMutationFn; data: LoginUserMutation | null | undefined },
  setters: {
    setInputUsernameSetter: typeof setInputUsername;
    setInputPasswordSetter: typeof setInputPassword;
    togglelIsLoadingSetter: typeof togglelIsLoading;
    setAlertSetter: typeof setAlert;
  },
  firstChatId: string,
) => {
  const { setInputUsernameSetter, setAlertSetter, setInputPasswordSetter, togglelIsLoadingSetter } = setters;
  const { data, loginUserMutation } = mutation;
  const rememberMe = auth.remember_me ? 31556952 : 3600;
  const refreshRememberMe = auth.remember_me ? 63113904 : 7200;
  togglelIsLoadingSetter(true);
  try {
    const { username } = values;
    await loginUserMutation({ variables: { ...values } });
    const { loginUser } = data || {};

    if (loginUser?.__typename === 'Error') setAlertSetter(loginUser?.message, 'error');
    else if (loginUser?.__typename === 'LoginUser') {
      const { AccessToken, RefreshToken, userId } = loginUser || {};

      const cookies = [
        { name: 'name', value: username, options: { sameSite: 'strict', maxAge: rememberMe, path: '/' } },
        { name: 'id', value: userId, options: { sameSite: 'strict', maxAge: rememberMe, path: '/' } },
        { name: 'token', value: AccessToken, options: { sameSite: 'strict', maxAge: rememberMe, path: '/' } },
        { name: 'refresh_token', value: RefreshToken, options: { sameSite: 'strict', maxAge: refreshRememberMe, path: '/' } },
      ];

      cookies.forEach((element) => {
        const { name, value, options } = element;
        cookie.set(name, value, { ...(options as any) });
      });

      cookie.set('REDIRECT_URL_CALLBACK', callback || `/${firstChatId}`);
      router.push(callback || `/${firstChatId}`);

      setInputUsernameSetter('');
      setInputPasswordSetter('');
    }
  } catch (error) {
    return error;
  } finally {
    togglelIsLoadingSetter(false);
  }
};
