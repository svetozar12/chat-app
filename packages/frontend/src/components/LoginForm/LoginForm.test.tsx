import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import LoginForm from "./LoginForm";
import renderer from "react-test-renderer";
import { AuthState } from "../../redux/reducer/authReducer";
import { render, screen } from "@testing-library/react";
import { initialState as inputState } from "../../redux/reducer/save_inputReducer";
import "@testing-library/jest-dom";

const submit = jest.fn();
submit.mockReturnValue("default");
const mockStore = configureStore([]);
const store = mockStore({
  authReducer: AuthState,
  saveInputReducer: inputState,
});

it("should create snapshot for <LoginForm/>", () => {
  expect(
    renderer
      .create(
        <Provider store={store}>
          <LoginForm handleSubmit={submit} />
        </Provider>,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it("should render <LoginForm/>", () => {
  render(
    <Provider store={store}>
      <LoginForm handleSubmit={submit} />
    </Provider>,
  );
  const renderedComponent = screen.getByText("Log In");
  expect(renderedComponent).toBeInTheDocument();
});
