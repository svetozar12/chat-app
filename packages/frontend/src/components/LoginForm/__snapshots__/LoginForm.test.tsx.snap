// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create snapshot for <LoginForm/> 1`] = `
<div
  className="flex"
  style={
    Object {
      "flexDirection": "column",
      "height": "100vh",
    }
  }
>
  
  <div
    style={
      Object {
        "height": "3rem",
        "width": "100%",
      }
    }
  />
  <h1
    className="css-my1y1y"
  >
    Login
  </h1>
  <form
    className="css-7rx0xo"
  >
    <div
      className="css-artkr0"
    >
      <label>
        Username
      </label>
      <input
        className="css-1s11o3x"
        name="username"
        onChange={[Function]}
        placeholder="username ..."
        type="text"
        value=""
      />
    </div>
    <div
      className="css-artkr0"
    >
      <label>
        Password
      </label>
      <input
        className="css-1s11o3x"
        name="password"
        onChange={[Function]}
        placeholder="password ..."
        type="password"
        value=""
      />
    </div>
    <button
      className="css-1hcgzlw"
      onClick={[MockFunction]}
      type="submit"
    >
      Log In
    </button>
    <div
      className="css-1sg2lsz"
    >
      <a
        className="link css-c9afss"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={
          Object {
            "color": "var(--button-blue)",
          }
        }
      >
        Sign up for chatApp .
      </a>
      <input
        checked={false}
        className="css-18woaps"
        data-testid="checkbox"
        id="checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="checkbox"
        style={
          Object {
            "marginLeft": ".25rem",
          }
        }
      >
         
        Remember me
      </label>
    </div>
  </form>
</div>
`;
